File: 3/js/p2.js
const readline = require('readline');

let binaries = [];

const rl = readline.createInterface({
  input: process.stdin
});

const prune = (bins, pos, tb) => {
  let count = 0;
  let ch = '';

  for (let i = 0; i < bins.length; i++) {
    const n = bins[i];

    if (n[pos] === '1') count++;
    else count--;
  }

  if (tb === 1) {
    if (count >= 0) ch = '1';
    else ch = '0';
  } else {
    if (count >= 0) ch = '0';
    else ch = '1';
  }

  const filtered = bins.filter(n => n[pos] === ch);

  if (filtered.length === 1) return filtered[0];
  else return prune(filtered, pos + 1, tb);
}

rl.on('line', (l) => {
  binaries.push(l);
});

rl.on('close', () => {
  const oxygen = prune(binaries, 0, 1);
  const co2 = prune(binaries, 0, 0);

  const on = parseInt(oxygen, 2);
  const cn = parseInt(co2, 2);

  console.log(`Oxygen: ${oxygen}, ${on}`);
  console.log(`CO2: ${co2}, ${cn}`);
  console.log(`${on * cn}`);
});

File: 3/lua/p2.lua
local function filter(t, f)
  local nt = {}

  for _,v in pairs(t) do
    if f(v) then table.insert(nt, v) end
  end

  return nt
end

local function prune(bins, pos, tb)
  local count = 0
  local ch = ''

  for _,n in ipairs(bins) do
    if n:sub(pos, pos) == '1' then
      count = count + 1
    else
      count = count - 1
    end
  end

  if tb == 1 then
    if count >= 0 then
      ch = '1'
    else
      ch = '0'
    end
  else
    if count >= 0 then
      ch = '0'
    else
      ch = '1'
    end
  end

  local filtered = filter(bins, function (n)
    return n:sub(pos, pos) == ch
  end)

  if #filtered == 1 then
    return filtered[1]
  else
    return prune(filtered, pos + 1, tb)
  end
end

local binaries = {}

for line in io.lines() do
  table.insert(binaries, line)
end

local oxygen = prune(binaries, 1, 1)
local co2 = prune(binaries, 1, 0)

print(oxygen, tonumber(oxygen, 2))
print(co2, tonumber(co2, 2))
print(tonumber(oxygen, 2) * tonumber(co2, 2))
